### ISSUE TITLE
Issues with complex on OSX Mavericks / libc++

### ISSUE THREAD

**Issue by tammojan** [casacore/casacore#85]
Possibly linked to this issue: https://llvm.org/svn/llvm-project/cfe/tags/Apple/clang-108.1/src/tools/clang/www/cxx_compatibility.html#dep_lookup

Also, note that since OSX 10.10 the standard is libc++ instead of libstdc++. A possible fix is to compile with -stdlib=libstdc++ but that may give problems when linking with programs that were built with the OS default libc++.

What steps will reproduce the problem?
1. Download casacore, install dependencies (particularly gcc for fortran) with brew
2. `cmake ../src/casacore`  (note that the C compiler is LLVM, and standard library is libc++ instead of libstdc++)
3. `make`

Is does not compile, output is:
```
Tammos-Mac-mini:build tammo$ make
[  0%] Building CXX object casa/CMakeFiles/casa_casa.dir/Arrays/ArrayOpsDiffShapes.cc.o
In file included from /Users/tammo/opt/casacore/trunk/src/casacore/casa/Arrays/ArrayOpsDiffShapes.cc:27:
In file included from /Users/tammo/opt/casacore/trunk/src/casacore/../casacore/casa/Arrays/ArrayOpsDiffShapes.h:32:
In file included from /Users/tammo/opt/casacore/trunk/src/casacore/../casacore/casa/Arrays/ArrayMath.h:33:
In file included from /Users/tammo/opt/casacore/trunk/src/casacore/../casacore/casa/BasicMath/Functors.h:33:
/Users/tammo/opt/casacore/trunk/src/casacore/../casacore/casa/BasicSL/Complex.h:177:16: error: invalid operands to binary expression
      ('const Complex' (aka 'const complex<Float>') and 'const Complex')
  { return left==right  ?  True : norm(left) >= norm(right); }
           ~~~~^ ~~~~~
/Users/tammo/opt/casacore/trunk/src/casacore/../casacore/casa/BasicSL/Complex.h:177:35: error: no matching function for call to
      'norm'
  { return left==right  ?  True : norm(left) >= norm(right); }
                                  ^~~~
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/complex:947:1: note: candidate function not viable: cannot convert
      argument of incomplete type 'const Complex' (aka 'const complex<Float>') to 'long double'
norm(long double __re)
^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/complex:954:1: note: candidate function not viable: cannot convert
      argument of incomplete type 'const Complex' (aka 'const complex<Float>') to 'double'
norm(double __re)
^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/complex:973:1: note: candidate function not viable: cannot convert
      argument of incomplete type 'const Complex' (aka 'const complex<Float>') to 'float'
norm(float __re)
^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/complex:936:1: note: candidate template ignored: could not match
      'std::__1::complex' against 'std::complex'
norm(const complex<_Tp>& __c)
^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/complex:963:5: note: candidate template ignored: disabled by
      'enable_if' [with _Tp = std::complex<float>]
    is_integral<_Tp>::value,
    ^
```

**Comment by gervandiepen** [comment]
Does it compile fine when using std::norm instead of norm?

**Comment by tammojan** [comment]
No:
```
/Users/tammo/opt/casacore/trunk/src/casacore/../casacore/casa/BasicSL/Complex.h:177:54: error: no matching function
      for call to 'norm'
  { return left==right  ?  True : std::norm(left) >= std::norm(right); }
                                                     ^~~~~~~~~
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/complex:947:1: note: candidate function not viable:
      cannot convert argument of incomplete type 'const Complex' (aka 'const complex<Float>') to 'long double'
norm(long double __re)
^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/complex:954:1: note: candidate function not viable:
      cannot convert argument of incomplete type 'const Complex' (aka 'const complex<Float>') to 'double'
norm(double __re)
^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/complex:973:1: note: candidate function not viable:
      cannot convert argument of incomplete type 'const Complex' (aka 'const complex<Float>') to 'float'
norm(float __re)
^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/complex:936:1: note: candidate template ignored: could
      not match 'std::__1::complex' against 'std::complex'
norm(const complex<_Tp>& __c)
^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/complex:963:5: note: candidate template ignored:
      disabled by 'enable_if' [with _Tp = std::complex<float>]
    is_integral<_Tp>::value,
```

**Comment by tammojan** [comment]
Fixed by Ger, making by changing in Complexfwd.h :
```
#ifdef AIPS_CXX11
#include <complex>
#else
namespace std {
  template<typename T> struct complex;
}
#endif
```

By simply
```
#include <complex>
```

**Comment by jtlz2** [comment]
@tammojan My Complexfwd.h doesn't look like yours, so not sure where to make that fix (casacore-1..5.0).

If you make that change, does pyrap subsequently install OK?

@ludwigschwardt What change did you make to https://github.com/ska-sa/homebrew-tap/blob/master/casacore.rb recently (since you had it fixed before)?

**Comment by ludwigschwardt** [comment]
Good news!

I've fixed casacore + pyrap last week for the impending casacore 2.0 release.

Just do:

`brew install casacore --devel`
`brew install pyrap`

**Comment by ludwigschwardt** [comment]
The trick is to compile with C++11 support, hence the `#ifdef AIPS_CXX11` part mentioned by @tammojan gets used - no code patching required.

**Comment by jtlz2** [comment]
@ludwigschwardt Wow - fantastic! What am I doing wrong? :-

> brew tap ska-sa/tap
Warning: Already tapped!

> brew install casacore --devel
Error: No devel block is defined for casacore

**Comment by ludwigschwardt** [comment]
brew update...

**Comment by gijzelaerr** [comment]
so this issue can be closed?

**Comment by jtlz2** [comment]
Not yet for me...

**Comment by ludwigschwardt** [comment]
Hi @jtlz2, try again like so (of course after updating brew):

```
brew install casacore
brew install pyrap
```

This gives me casacore 2.0.1 and pyrap 1.1.0. Both give a warning about gcc being built with a different C++ standard library but I ignore that.

I can open MS tables in Python like so:

```
from pyrap import tables
t = tables.table('CygA-X.MS', readonly=False, ack=False)
t.colnames()
```

Let me know how your experience differs.

**Comment by jtlz2** [comment]
Thanks @ludwigschwardt - here you go:

cordelia:~ jtlz2$ brew install casacore --verbose
==> Installing casacore from ska-sa/homebrew-tap
==> Using Homebrew-provided fortran compiler.
This may be changed by setting the FC environment variable.
==> Downloading https://github.com/casacore/casacore/archive/v2.0.1.tar.gz
Already downloaded: /Users/jtlz2/Library/Caches/Homebrew/casacore-2.0.1.tar.gz
==> Verifying casacore-2.0.1.tar.gz checksum
tar xf /Users/jtlz2/Library/Caches/Homebrew/casacore-2.0.1.tar.gz
==> cmake ../.. -DCMAKE_C_FLAGS_RELEASE= -DCMAKE_CXX_FLAGS_RELEASE= -DCMAKE_INSTALL_PREFIX=/usr/local/Cellar/casacore/2.0.1 -DCMAKE_BUILD_TYPE=Release -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_VERBOSE_MAKEFILE=ON -Wno-dev -DCMAKE_BUILD_TYPE=release -DBUILD_PYTHON=ON -DUSE_FFTW3=ON -DFFTW3_ROOT_DIR=/usr/local -DUSE_HDF5=ON -DHDF5_ROOT_DIR=/usr/local -DUSE_THREADS=ON -DDATA_DIR=/usr/local/share/casacore/data
-- The C compiler identification is AppleClang 6.0.0.6000057
-- The CXX compiler identification is AppleClang 6.0.0.6000057
-- Check for working C compiler: /usr/local/Library/ENV/4.3/clang
-- Check for working C compiler: /usr/local/Library/ENV/4.3/clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/local/Library/ENV/4.3/clang++
-- Check for working CXX compiler: /usr/local/Library/ENV/4.3/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
dpkg: warning, architecture `x86_64-apple-darwin13.4.0' not in remapping table
-- The Fortran compiler identification is GNU
-- Checking whether Fortran compiler has -isysroot
-- Checking whether Fortran compiler has -isysroot - yes
-- Checking whether Fortran compiler supports OSX deployment target flag
-- Checking whether Fortran compiler supports OSX deployment target flag - yes
-- Check for working Fortran compiler: /usr/local/bin/gfortran
-- Check for working Fortran compiler: /usr/local/bin/gfortran  -- broken
CMake Error at /usr/local/Cellar/cmake/3.2.2/share/cmake/Modules/CMakeTestFortranCompiler.cmake:54 (message):
  The Fortran compiler "/usr/local/bin/gfortran" is not able to compile a
  simple test program.

  It fails with the following output:

   Change Dir: /tmp/casacore20150504-13255-w2qefn/casacore-2.0.1/build/release/CMakeFiles/CMakeTmp

  

  Run Build Command:"/usr/local/Library/ENV/4.3/gmake"
  "cmTryCompileExec4032099131/fast"

  /Applications/Xcode.app/Contents/Developer/usr/bin/make -f
  CMakeFiles/cmTryCompileExec4032099131.dir/build.make
  CMakeFiles/cmTryCompileExec4032099131.dir/build

  /usr/local/Cellar/cmake/3.2.2/bin/cmake -E cmake_progress_report
  /tmp/casacore20150504-13255-w2qefn/casacore-2.0.1/build/release/CMakeFiles/CMakeTmp/CMakeFiles
  1

  Building Fortran object
  CMakeFiles/cmTryCompileExec4032099131.dir/testFortranCompiler.f.o

  /usr/local/bin/gfortran -c
  /tmp/casacore20150504-13255-w2qefn/casacore-2.0.1/build/release/CMakeFiles/CMakeTmp/testFortranCompiler.f
  -o CMakeFiles/cmTryCompileExec4032099131.dir/testFortranCompiler.f.o

  Linking Fortran executable cmTryCompileExec4032099131

  /usr/local/Cellar/cmake/3.2.2/bin/cmake -E cmake_link_script
  CMakeFiles/cmTryCompileExec4032099131.dir/link.txt --verbose=1

  /usr/local/bin/gfortran
  CMakeFiles/cmTryCompileExec4032099131.dir/testFortranCompiler.f.o -o
  cmTryCompileExec4032099131

  Undefined symbols for architecture x86_64:

    "__gfortran_set_options", referenced from:
        _main in testFortranCompiler.f.o
    "__gfortran_transfer_character_write", referenced from:
        _MAIN__ in testFortranCompiler.f.o

  ld: symbol(s) not found for architecture x86_64

  collect2: error: ld returned 1 exit status

  make[1]: *** [cmTryCompileExec4032099131] Error 1

  make: *** [cmTryCompileExec4032099131/fast] Error 2

  

  

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:45 (enable_language)


-- Configuring incomplete, errors occurred!
See also "/tmp/casacore20150504-13255-w2qefn/casacore-2.0.1/build/release/CMakeFiles/CMakeOutput.log".
See also "/tmp/casacore20150504-13255-w2qefn/casacore-2.0.1/build/release/CMakeFiles/CMakeError.log".
Warning: It appears you have MacPorts or Fink installed.
Software installed with other package managers causes known problems for
Homebrew. If a formula fails to build, uninstall MacPorts/Fink and try again.
==> Formula
Tap: ska-sa/homebrew-tap
Path: /usr/local/Library/Taps/ska-sa/homebrew-tap/casacore.rb
==> Configuration
HOMEBREW_VERSION: 0.9.5
ORIGIN: https://github.com/Homebrew/homebrew
HEAD: 6cb91e45713cb1c622cb54999116686f369c5b2f
Last commit: 9 hours ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CELLAR: /usr/local/Cellar
CPU: 8-core 64-bit haswell
OS X: 10.9.5-x86_64
Xcode: 6.2
CLT: 6.2.0.0.1.1424975374
GCC-4.2: build 5666
Clang: 6.0 build 600
MacPorts/Fink: /opt/local/bin/port, /sw/bin/fink
X11: 2.7.7 => /opt/X11
System Ruby: 2.0.0-p481
Perl: /opt/local/bin/perl => /opt/local/bin/perl5.16
Python: /Users/jtlz2/Library/Enthought/Canopy_64bit/User/bin/python
Ruby: /usr/bin/ruby
Java: 1.6.0_65
==> ENV
HOMEBREW_CC: clang
HOMEBREW_CXX: clang++
MAKEFLAGS: -j8
CMAKE_PREFIX_PATH: /usr/local/opt/readline:/usr/local
CMAKE_INCLUDE_PATH: /usr/include/libxml2:/System/Library/Frameworks/OpenGL.framework/Versions/Current/Headers
CMAKE_LIBRARY_PATH: /System/Library/Frameworks/OpenGL.framework/Versions/Current/Libraries
PKG_CONFIG_PATH: /usr/local/opt/cfitsio/lib/pkgconfig:/usr/local/opt/wcslib/lib/pkgconfig:/usr/local/opt/fftw/lib/pkgconfig
PKG_CONFIG_LIBDIR: /usr/lib/pkgconfig:/usr/local/Library/ENV/pkgconfig/10.9
ACLOCAL_PATH: /usr/local/share/aclocal
PATH: /usr/local/Library/ENV/4.3:/usr/local/opt/cmake/bin:/usr/local/opt/wcslib/bin:/usr/local/opt/fftw/bin:/usr/local/opt/hdf5/bin:/usr/local/opt/gcc/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin

Error: casacore 2.0.1 did not build
Logs:
     /Users/jtlz2/Library/Logs/Homebrew/casacore/01.cmake
     /Users/jtlz2/Library/Logs/Homebrew/casacore/01.cmake.cc
     /Users/jtlz2/Library/Logs/Homebrew/casacore/CMakeCache.txt

**Comment by ludwigschwardt** [comment]
Your gfortran is borked...

Another user had [a similar issue](http://www.open-mpi.org/community/lists/users/2013/12/23110.php) with Open MPI and this was his final resolution:

> If any Homebrew users have this issue in the future when installing 
> open-mpi here's what happened: there were some non-Homebrewed 32-bit 
> gfortran libraries floating around in the lib directory Homebrew uses,
> which were being called instead of the correct Homebrewed 64 bit libraries. 

Maybe use `brew doctor` to find and remove any culprits.

**Comment by tammojan** [comment]
This issue was fixed with commit 4483c124a35e820d00e1cc267732781a49b6aec1
